Migrate spezi health kit

max-rosenblattl opened the pull request:
# Migrate HealthKit to SpeziHealthKit

## :recycle: Current situation & Problem
The current implementation doesn't take advantage of `SpeziHealthKit`.


## :gear: Release Notes
This PR migrates from HealthKit to SpeziHealthKit, improving readability and simplifying the access of health data. 


## :books: Documentation
* This PR mainly refactors the existing implementation
* The added extension of `HealthQuery` should optimally be merged into the `SpeziHealthKit` framework.


## :white_check_mark: Testing
Tested on iPhone Air.
| <img src="https://github.com/user-attachments/assets/e93d0312-3fbf-4b7c-90cc-c95b96a9f39a" alt="IMG_0126" width="200"> | <img src="https://github.com/user-attachments/assets/8269a454-6deb-44a1-9f96-16173cd280a4" alt="IMG_0127" width="200"> | <img src="https://github.com/user-attachments/assets/b38f8b59-93f9-49ce-8626-02d8143a11bb" alt="IMG_0129" width="200"> |
|:-------------------------------------------------------:|:-------------------------------------------------------:|:-------------------------------------------------------:|
| **Workout Info 1** | **Workout Info 2** | **Health Data** |


| <img width="630" height="1368" alt="IMG_0136" src="https://github.com/user-attachments/assets/3debb426-c84c-4d5b-9931-ab02a2a96d40" /> | <img width="630" height="1368" alt="IMG_0137" src="https://github.com/user-attachments/assets/e6bb694f-0cb3-418a-912c-02b4cb2658af" /> | <img width="630" height="1368" alt="IMG_0138" src="https://github.com/user-attachments/assets/f6a1e3f8-da46-44e1-90be-c59ec34dcd69" /> |
|:-------------------------------------------------------:|:-------------------------------------------------------:|:-------------------------------------------------------:|
| **Time Range Agg. 1** | **Time Range Agg. 2** | **Time Range Agg. 3** |





### Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

max-rosenblattl committed:
Add basic
SpeziHealthKit support

max-rosenblattl committed:
Adapt health data fetcher for aggregation

max-rosenblattl committed:
Add aggregation support for SpeziHealthKit query

max-rosenblattl committed:
Refactor fetch health

max-rosenblattl committed:
Add parallel processing to fetch health

max-rosenblattl committed:
Refactor user info

max-rosenblattl committed:
Fix health characteristic access

max-rosenblattl committed:
Refactor fetch workout

max-rosenblattl committed:
Remove collect samples

max-rosenblattl committed:
Refactor health kit authorization

max-rosenblattl committed:
Fix statistics query bug

max-rosenblattl committed:
Revert debugging tools

max-rosenblattl commented on HealthyLLM.xcodeproj/project.pbxproj:
```diff
@@ -534,7 +544,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = 1;
 				DEVELOPMENT_ASSET_PATHS = "\"HealthyLLM/Supporting Files/Preview Content\"";
-				DEVELOPMENT_TEAM = CQRZ4E7K9U;
```
I will revert the development team and bundle identifier after addressing all other PR comments

max-rosenblattl commented on HealthyLLM/HealthyLLM/Fetcher/HealthDataFetcher.swift:
```diff
@@ -8,99 +8,191 @@
         return result
     }
 }
+
+// One-off query for aggregating health data
+// TODO: Create Pull Request for SpeziHealthKit
+extension HealthKit {
+    public func statisticsQuery<Sample>(
```
@lukaskollmer do you know if `SpeziHealthKit` already provides the functionality for one-off statistics queries in some way? If not, and you think it would be useful to others as well, I'm happy to create a PR!

lukaskollmer replied:
i don't think so; iirc i was planning to create a nice API around HKStatistics (eg so that you dont have to deal with the optional HKQuantity objects that will never be nil) but i didn't get around to it. we do have the `@HealthKitStatisticsQuery` for SwiftUI; would make sense to add a non-SwiftUI counterpart for one-off queries. we might even be able to reuse the code internally

lukaskollmer replied:
the `@HealthKitStatisticsQuery` API already contains an initial attempt at making this type safe; we have different init overloads to prevent the caller from passing in invalid aggregation option combinations

PSchmiedmayer replied:
@lukaskollmer and I had a small discussion on Monday about a very simple approach to make the SwiftUI property wrappers also available outside of the SwiftUI hierarchy and usable inside modules, we might want to tackle this in the future, IMO that would solve the issue here as well. But we might want merge this first and then iterate on it later.

max-rosenblattl replied:
Sounds good, @lukaskollmer let me know if you need some support, but for now I agree with merging first and iterating later 👍

PSchmiedmayer replied:
@max-rosenblattl Would be great if you could create an issue in SpeziHealthKit that tracks this issue. I also saw that @phnagy was looking into some issues there; maybe that might be a great first issue with a real need and direct implementation pathway. (CC @lukaskollmer)

max-rosenblattl replied:
@lukaskollmer I phrased the [issue](https://github.com/StanfordSpezi/SpeziHealthKit/issues/64) in a way that suggests implementing your discussed approach; if you have any, it would be great to add some specific details from your discussion that could ease the implementation.

PSchmiedmayer replied:
Amazing; thank you @max-rosenblattl!

max-rosenblattl commented on HealthyLLM/HealthyLLM/Fetcher/HealthDataFetcher.swift:
```diff
@@ -8,99 +8,191 @@
-                taskGroup.addTask {
-                    (
-                        description,
-                        try? await self.fetchSample(
-                            for: identifier,
-                            unit: unit,
-                            startDate: endDate,
-                            endDate: .now
-                        )
```
Currently, we’re aggregating time-series data into a single value (e.g., average heart rate over the last month). It might be more meaningful to use a different aggregation approach; for example, retrieving daily averages over the last month instead of just one number. This would also align better with future EmbedHealth integration, even if that’s a larger effort. Let me know what you think @PSchmiedmayer @LeonNissen @lukaskollmer

lukaskollmer replied:
agree that a higher granularity would be desirable but wasn't the issue here that having too many samples could exceed the context window? (i might be confusing this with something else).

iirc another idea of mine at the time was that we could use different granularities depending on how far back the data was collected (eg weekly heart rate for > a year ago, daily average for > last week, and hourly averages for the past 7 days)

PSchmiedmayer replied:
Yes, the main issue was the growing context window; that's why we constrained it to a few samples.

Agree with @lukaskollmer here for this approach. I think 7 days for a week, weekly samples for a month, or monthly samples for a year would make sense and don't dramatically exceed the context window. Would be good to verify this with some on-device testing @max-rosenblattl before we merge this.

max-rosenblattl replied:
Thanks for the insights! I think it makes a lot of sense, happy to tackle that. 

Regarding testing on device: I will be able to do so in the very near future; until then if anyone has some spare minutes, I would appreciate it!

max-rosenblattl replied:
Quick update: finally got my new iPhone and could test it on device (results in PR description above)! Also adapted the statistics query to handle the discussed time-range based aggregation intervals. Just saw they were integrated in here too https://github.com/StanfordSpezi/SpeziHealthKit/pull/65 , quite cool looking forward to the merge 🚀

PSchmiedmayer replied:
Thank you for working on all of this @max-rosenblattl 🚀

PSchmiedmayer reviewed (commented):
Thank you for working on this and thank you for your initial comments @lukaskollmer 👍

Happy to see a first version merged once the CI is passing; we will have a decent amount of iterations here when we will embed the time series LLM components but that is a next step after that.

max-rosenblattl committed:
Implement time range based aggregation of health records

max-rosenblattl committed:
Fix SwiftLint violations and license headers (#4)

# Fix Swfitlint errors

## :recycle: Current situation & Problem
Currently, the PR analyzers don't pass because of multiple code
analyzers failing.


## :gear: Release Notes
This PR fixes Swiftlint errors and warnings.


## :books: Documentation
Only refactoring, no documentation needed.


## :white_check_mark: Testing
*Please ensure that the PR meets the testing requirements set by CodeCov
and that new functionality is appropriately tested.*
*This section describes important information about the tests and why
some elements might not be testable.*


### Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our
[Code of
Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of
Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md)
and [Contributing
Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

max-rosenblattl committed:
Fix unused variables

max-rosenblattl committed:
Refactor statistics query using SpeziHealthKit

max-rosenblattl committed:
Fix SwiftLint warning
